### PR TITLE
feature widget description

### PR DESCRIPTION
**Pull request by CamilleLegeron**
#431
Peek 2023-04-17 10-46

### CODE REVIEW

**Comment by CamilleLegeron** [comment]
Hello @berhalak 
I've done the widget description but there might be some work to do to fit the style of he tooltip you've done [here](https://github.com/gristlabs/grist-core/pull/441)

**Comment by berhalak** [comment]
> Hello @berhalak I've done the widget description but there might be some work to do to fit the style of he tooltip you've done [here](https://github.com/gristlabs/grist-core/pull/441)

Grid tooltip and the new rename dialog were merged and will be available on Monday.

**Comment by CamilleLegeron** [comment]
Thanks @berhalak for the work,
I adapt my PR, you can review it now

File: .gitignore
Comment (georgegevoian) [inline .gitignore:1]: These files are already ignored on lines 78-79.

File: app/client/ui/DescriptionConfig.ts [hunk 1]
```diff
@@ -11,8 +11,9 @@ const t = makeT('FieldConfig');
 
 export function buildDescriptionConfig(
     owner: MultiHolder,
-    origColumn: ColumnRec,
+    parent: ColumnRec | ViewSectionRec,
```
Comment (georgegevoian) [inline app/client/ui/DescriptionConfig.ts:14]:
```suggestion
    description: KoSaveableObservable<string>,
```

Since we really only need the `description` here, we can change the parameter so that the callers have to pass in the description.

File: app/client/ui/DescriptionConfig.ts [hunk 2]
```diff
@@ -11,8 +11,9 @@ const t = makeT('FieldConfig');
...
     cursor: ko.Computed<CursorPos>,
+    testPrefix: string,
```
Comment (georgegevoian) [inline app/client/ui/DescriptionConfig.ts:16]: Sorry for nitpicking, but can you wrap these two parameters into a single `options` parameter?

```suggestion
    options: {
      cursor: ko.Computed<CursorPos>,
      testPrefix: string,
    }    
```

**Review by georgegevoian — CHANGES_REQUESTED** [review]
Thanks @CamilleLegeron. Tried it out and it's working well.

Two things I noticed:
1. The widget description textarea in the popup has an outline when focused. Since the other description textareas don't have one, we should probably remove it for consistency.
2. The widget description textarea in the popup doesn't add a newline when Enter/Return is pressed, which is the behavior in the column description textarea. If I recall correctly, either Ctrl+Enter or Mod+Enter overrides the newline behavior and saves instead. I think it'd be good to keep the behavior consistent and match that how column descriptions work.

**Comment by CamilleLegeron** [comment]
Thanks @georgegevoian for all your corrections. I took them into account and adapted the tests

File: app/client/ui/DescriptionConfig.ts [hunk 3]
```diff
@@ -23,22 +25,22 @@ export function buildDescriptionConfig(
     // update a different column.
     let editor: HTMLTextAreaElement | undefined;
     owner.autoDispose(
-      cursor.subscribe(() => {
+      options.cursor?.subscribe(() => {
```
Comment (georgegevoian) [inline app/client/ui/DescriptionConfig.ts:28]:
```suggestion
      options.cursor.subscribe(() => {
```

File: app/client/ui/WidgetTitle.ts [hunk 4]
```diff
@@ -144,30 +222,41 @@ function buildWidgetRenamePopup(ctrl: IOpenController, vs: ViewSectionRec, optio
...
     dom.maybe(!options.widgetNameHidden, () => [
       cssLabel(t("WIDGET TITLE")),
       widgetInput = cssInput(inputWidgetTitle, updateOnKey, {placeholder: inputWidgetPlaceholder},
-        testId('section-name-input')
+        testId('section-name-input'),
+        commandGroup.attach(),
       ),
     ]),
+    cssLabel(t("WIDGET DESCRIPTION")),
+    descInput = cssTextArea(inputWidgetDesc, updateOnKey,
```
Comment (georgegevoian) [inline app/client/ui/WidgetTitle.ts:237]: What do you think about moving the `cssTextArea` out of `ColumnTitle.ts` and into a new file, `DescriptionTextArea.ts`, and exporting a function from it that creates it, taking optional dom arguments as a parameter?

Both `WidgetTitle.ts` and `ColumnTitle.ts` could then use it, passing in their own command groups and test ids.

File: app/client/ui/WidgetTitle.ts [hunk 5]
```diff
@@ -122,18 +149,69 @@ function buildWidgetRenamePopup(ctrl: IOpenController, vs: ViewSectionRec, optio
...
+  // Create this group and attach it to the popup and all inputs.
+  const commandGroup = commands.createGroup({ ...myCommands }, ctrl, true);
+
   let tableInput: HTMLInputElement|undefined;
   let widgetInput: HTMLInputElement|undefined;
+  let descInput: HTMLTextAreaElement | undefined;
   return cssRenamePopup(
-    // Create a FocusLayer to keep focus in this popup while it's active, and prevent keyboard
-    // shortcuts from being seen by the view underneath.
-    elem => { FocusLayer.create(ctrl, {defaultFocusElem: elem, pauseMousetrap: true}); },
```
Comment (georgegevoian) [inline app/client/ui/WidgetTitle.ts:136]: We should keep this and just change `pauseMousetrap` to `false`.

**Review by georgegevoian — CHANGES_REQUESTED** [review]

File: app/client/ui/DescriptionConfig.ts (hunk 3 above)
Comment (CamilleLegeron) [reply app/client/ui/DescriptionConfig.ts:28]: If we put the cursor in the `options` object it's because it's optional no ? Otherwise I don't understand why we would want to do this

Comment (georgegevoian) [reply app/client/ui/DescriptionConfig.ts:28]: It's not optional currently; if we don't pass in a cursor, we'll get a type error. It's ok for us to have required properties in `options` though.

The reason why I suggested the change is because we had too many parameters (more than 3).

File: app/client/ui/WidgetTitle.ts (hunk 4 above)
Comment (CamilleLegeron) [reply app/client/ui/WidgetTitle.ts:240]: Maybe we can create file with all common styles as RightPanelStyles, because it's the main part to factorize

**Comment by CamilleLegeron** [comment]
I don't understand why the test works fine and just with my last changes it breaks ... 
In local ChoiceList seems to be the cause but I don't know why : 
image
 Any idea ?

File: app/client/ui/WidgetTitle.ts (hunk 5 above)
Comment (CamilleLegeron) [reply app/client/ui/WidgetTitle.ts:136]: It's not marked as outdated but I changed it as you said

File: app/client/ui/WidgetTitle.ts [hunk 6]
```diff
@@ -210,12 +210,15 @@ function buildWidgetRenamePopup(ctrl: IOpenController, vs: ViewSectionRec, optio
...
+    // Create a FocusLayer to keep focus in this popup while it's active, and prevent keyboard
+    // shortcuts from being seen by the view underneath.
+    elem => { FocusLayer.create(ctrl, { defaultFocusElem: elem, pauseMousetrap: false }); },
     dom.onDispose(onClose),
     dom.autoDispose(commandGroup),
     testId('popup'),
     dom.cls(menuCssClass),
     dom.maybe(!options.tableNameHidden, () => [
-      cssLabel(t("DATA TABLE NAME")),
+      cssLabel(t("Data table name")),
```
Comment (georgegevoian) [inline app/client/ui/WidgetTitle.ts:221]: Why the change? I think this breaks existing translations.

**Review by georgegevoian — CHANGES_REQUESTED** [review]
Thanks for the updates @CamilleLegeron. Just have one more comment.

File: app/client/ui/WidgetTitle.ts (hunk 6 above)
Comment (CamilleLegeron) [reply app/client/ui/WidgetTitle.ts]: oooh you're right, I forgot that .. --'
I had changed it because I think it's cleaner to say it's uppercase in the css because it's UI decision, but better not to break existing translations

**Review by georgegevoian — APPROVED** [review]
Thanks @CamilleLegeron!
